track ignored threads and cancel them when Lwt_extra.run finishes

domsj opened the pull request:
CI still running

NicolasT commented:
See Skype-feedback about `run` implementation.

NicolasT commented:
If CI and @toolslive agree, LGTM.

domsj commented:
So far CI ain't agreeing, still making some changes :)

domsj committed:
track ignored threads and cancel them when Lwt_extra.run finishes

NicolasT commented on src/tools/lwt_extra.ml:
```diff
@@ -0,0 +1,69 @@
+
+let igns = Hashtbl.create 10;;
+
+let () =
```
Not sure having side-effects to be executed 'at module loading time' is a good idea (and seems unnecessary here, read on)

NicolasT commented on src/tools/lwt_extra.ml:
```diff
@@ -0,0 +1,69 @@
+let igns = Hashtbl.create 10;;
+
+let () =
+  Random.self_init ();
```
No need to init the PRNG. All we need is some pseudo-random numbers, no need to have some impure seed or whatever.

NicolasT commented on src/tools/lwt_extra.ml:
```diff
@@ -0,0 +1,69 @@
+
+let () =
+  Random.self_init ();
+  Lwt.async_exception_hook :=
```
I'd wrap setting and restoring `async_except_hook` in the `run` implementation, so normal Lwt behavior is kept.

domsj replied:
Ok

NicolasT commented on src/tools/lwt_extra.ml:
```diff
@@ -0,0 +1,69 @@
+  Random.self_init ();
+  Lwt.async_exception_hook :=
+    (function
+      | Lwt.Canceled -> ()
```
Maybe log something instead?

domsj replied:
I can't use Logger.bla here as these return unit Lwt.t
Also, these cancelleds are actually to be expected

NicolasT commented on src/tools/lwt_extra.ml:
```diff
@@ -0,0 +1,69 @@
+        Hashtbl.remove igns key;
+        Lwt.return ()) in
+  Hashtbl.add igns key t;
+  Lwt.ignore_result t';;
```
No need for ;;

NicolasT commented on src/tools/lwt_extra.ml:
```diff
@@ -0,0 +1,69 @@
+      (fun () ->
+        Hashtbl.remove igns key;
+        Lwt.return ()) in
+  Hashtbl.add igns key t;
```
Consider adding this _inside_ the action contained in `Lwt.finalize` (i.e. on line 47).

Rationale: what if `Lwt_main.run` returns after the key & action are added to the map, but before the next line is reached? They should only be in the map if the action has actually been launched.

domsj replied:
Ok, considered this.

I want to add the k,v pair to the hashtbl before other threads get the chance to run and possibly ignore another thread, for which the random number generator would return the same number (ok, chances for this are very very slim, but then maybe I should also remove the logic in get_key?)

As for the what if, the Lwt_main.run can only finish when it receives control again (e.g. when a Lwt.bind happens). This can only happen after the key has been added and the thread has successfully been ignored (and possibly even line 47 would at that time not yet be executed). So I'll leave this as is.

NicolasT replied:
OK

NicolasT commented on src/tools/lwt_extra.ml:
```diff
@@ -0,0 +1,69 @@
+      (fun () ->
+        Lwt.catch
+          (fun () ->
+            let ignored_threads = Hashtbl.fold (fun k t acc -> t :: acc) igns [] in
```
I'm surprised this doesn't generate an 'unused binding' compiler warning for `k`.

domsj replied:
Yea, sorta suprising. Maybe we need an extra compiler flag to trigger warnings for this?

NicolasT commented on src/tools/lwt_extra.ml:
```diff
@@ -0,0 +1,69 @@
+            Hashtbl.reset igns;
+            Lwt.pick (Lwt.return () :: ignored_threads) >>= fun a ->
+            Lwt.return ())
+          (fun exn -> Lwt.return ()))
```
Consider logging the exception?

domsj replied:
Ok

domsj committed:
processed feedback from pull request

domsj committed:
detached compression thread is canceled too when Lwt_extra.run ends
Lwt_extra.async_exception_hook should be used as of now instead of Lwt.async_exception_hook

NicolasT commented on src/tools/lwt_extra.ml:
```diff
@@ -0,0 +1,92 @@
+  Hashtbl.add igns key t;
+  Lwt.ignore_result (t' ())
+
+let detach (t: unit -> string) =
```
Is the type restriction required?

domsj replied:
As it is implemented right now, yes, otherwise the hashtbl above (https://github.com/Incubaid/arakoon/pull/90/files#L26R27) has an 'a in its type which the compiler complains about.
So for now only detached computations returning strings are supported. I think each return type needs it's own Hashtbl tracking the openstanding threads? Or maybe some other clever implementation :), will think if I can find a better way.

NicolasT commented on src/tools/lwt_extra.ml:
```diff
@@ -0,0 +1,92 @@
+      Hashtbl.remove detacheds key;
+      Lwt.return ())
+
+let async_exception_hook : (exn -> unit) ref =
```
I don't understand why a ref is used here.

domsj replied:
It's the same as here: https://github.com/avsm/lwt/blob/master/src/core/lwt.ml#L179
The idea is that this can be set/configured from anywhere else statically.
Another option would be to pass in the async_exception_hook as a parameter of run.

NicolasT commented on src/tools/lwt_extra.ml:
```diff
@@ -0,0 +1,92 @@
+let () =
+  Lwt.async_exception_hook :=
+    (function
+      | Lwt.Canceled -> ()
```
I'm (still) not convinced it's a good idea to have this as the default. `Lwt.Canceled` in threads during normal operation is a bug. Only when `run` finalizes, these things are expected to occur.
